sdk/openai/azure-ai-openai-assistants/src/main/java/com/azure/ai/openai/assistants/implementation/streaming/OpenAIServerSentEvents.java

```diff
@@ -0,0 +1,178 @@
+package com.azure.ai.openai.assistants.implementation.streaming;
+
+import com.azure.ai.openai.assistants.implementation.models.AssistantStreamEvent;
+import com.azure.ai.openai.assistants.models.StreamUpdate;
+import com.azure.core.util.BinaryData;
+import reactor.core.publisher.Flux;
+import reactor.core.scheduler.Schedulers;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.azure.ai.openai.assistants.implementation.models.AssistantStreamEvent.DONE;
+import static com.azure.ai.openai.assistants.implementation.models.AssistantStreamEvent.ERROR;
+
+/**
+ * A class that handles the deserialization of server sent events.
+ */
+public final class OpenAIServerSentEvents {
+
+    // Server sent events are divided by 2 CRLF or single LF character
+    private static final int SSE_CHUNK_LINE_BREAK_COUNT_MARKER = 2;
+
+    /**
+     * A factory that determines into which type to deserialize the server sent events.
+     */
+    private final StreamTypeFactory eventDeserializer = new StreamTypeFactory();
+
+    /**
+     * The source of the server sent events.
+     */
+    private final Flux<ByteBuffer> source;
+
+    /**
+     * The output stream accumulating the server sent events.
+     */
+    private ByteArrayOutputStream outStream;
+
+    /**
+     * Creates a new instance of OpenAIServerSentEvents.
+     *
+     * @param source The source of the server sent events.
+     */
+    public OpenAIServerSentEvents(Flux<ByteBuffer> source) {
+        this.source = source;
+        this.outStream = new ByteArrayOutputStream();
+    }
+
+    /**
+     * Gets the stream of server sent events.
+     *
+     * @return A stream of server sent events.
+     */
+    public Flux<StreamUpdate> getEvents() {
```

Comment:
Will we always return `StreamUpdate` type or should we make this type generic like we have for [azure-ai-openai](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/implementation/OpenAIServerSentEvents.java#L20)

Reply:
I believe we always want to return `StreamUpdate`. In inference you always had some version of `ChatCompletions` (which had to have almost everything nullable because of this) emitted. For assistants, depending on the event emitted by the backend you have an entirely different class emitted. Depending on the event, it will map to a different type, that's why I thought to replicate the approach in .NET to have a container type to deserialize accordingly.

Here is all the (currently) possible types we might have to deserialize into from [OpenAI](https://github.com/openai/openai-openapi/blob/b4d127c88bedb2572d69595e74a25918cd0d0316/openapi.yaml#L12329-L12359) and how they map into our [TSP definition](https://github.com/Azure/azure-rest-api-specs/blob/f92727a3b24f31bea56021b68aec02fb68a557f9/specification/ai/OpenAI.Assistants/streaming/events.tsp) (still in review)